### PR TITLE
s2571 - grant reader/writer, config parsing, GetAction

### PR DESCRIPTION
**Pull request by jburke-cadc**

### CODE REVIEW

File: cadc-storage-permissions/src/main/java/org/opencadc/inventory/permissions/xml/GrantReader.java [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+import org.jdom2.JDOMException;
+import org.opencadc.gms.GroupURI;
+import org.opencadc.inventory.permissions.Grant;
+import org.opencadc.inventory.permissions.ReadGrant;
+import org.opencadc.inventory.permissions.WriteGrant;
+
+public class GrantReader {
+
+    static enum ENAMES {
+        grant(), artifactURI(), expiryDate(), isAnonymousAccess(), groups(), groupURI(),
```
Comment (pdowler) [inline cadc-storage-permissions/src/main/java/org/opencadc/inventory/permissions/xml/GrantReader.java]: isAnonymousAccess -> anonymousRead or anonRead

File: cadc-storage-permissions/src/test/resources/valid.properties [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# Sample properties file for cadc-storage-permissions.
```
Comment (pdowler) [inline cadc-storage-permissions/src/test/resources/valid.properties]: Although this is what we talked about, now that I see these examples I think it is pretty messy in a few cases:

Any pattern with anon = T doesn't need any read-only groups so ops has to config a dummy group.
Anything beyond one group is easy to mess up syntactically... maybe we need something that allows null values so only positive (granting) config is required. TBD

File: baldur/src/main/webapp/WEB-INF/web.xml [hunk 3]
```diff
@@ -16,13 +16,14 @@
         <servlet-class>ca.nrc.cadc.log.LogControlServlet</servlet-class>
         <init-param>
             <param-name>logLevel</param-name>
-            <param-value>info</param-value>
+            <param-value>debug</param-value>
```
[the remark this answers is not in the document]
Comment (jburke-cadc) [reply baldur/src/main/webapp/WEB-INF/web.xml]: Done

File: cadc-storage-permissions/src/main/java/org/opencadc/inventory/permissions/GetAction.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        java.net.URI artifactURI;
+        try {
+            artifactURI = new URI(uri);
+        } catch (URISyntaxException e) {
+            throw new ServletException(URI + " parameter is not a valid URI");
+        }
+        log.debug(OP + " = " + op);
+        log.debug(URI + " = " + artifactURI.toASCIIString());
+
+        // Check if the calling user is authorized to make the request.
```
[the remark this answers is not in the document]
Comment (jburke-cadc) [reply cadc-storage-permissions/src/main/java/org/opencadc/inventory/permissions/GetAction.java]: Agree and changed.

File: cadc-storage-permissions/build.gradle [hunk 5]
```diff
@@ -21,6 +21,7 @@ dependencies {
     compile 'log4j:log4j:[1.2,)'
     compile 'org.opencadc:cadc-inventory:[0.1,)'
     compile 'org.opencadc:cadc-gms:[1.0,)'
+    compile 'org.opencadc:cadc-rest:[1.2.1,2.0)'
```
[the remark this answers is not in the document]
Comment (jburke-cadc) [reply cadc-storage-permissions/build.gradle]: Removed.

File: cadc-storage-permissions/src/main/java/org/opencadc/inventory/permissions/GetAction.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+ *  General Public License along         Publique GNU Affero avec
+ *  with OpenCADC.  If not, see          OpenCADC ; si ce n’est
+ *  <http://www.gnu.org/licenses/>.      pas le cas, consultez :
+ *                                       <http://www.gnu.org/licenses/>.
+ *
+ *  $Revision: 5 $
+ *
+ ************************************************************************
+ */
+
```
[the remark this answers is not in the document]
Comment (jburke-cadc) [reply cadc-storage-permissions/src/main/java/org/opencadc/inventory/permissions/GetAction.java]: Moved.

File: cadc-storage-permissions/src/main/java/org/opencadc/inventory/permissions/PermissionsAction.java [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+ *  with OpenCADC.  If not, see          OpenCADC ; si ce n’est
+ *  <http://www.gnu.org/licenses/>.      pas le cas, consultez :
+ *                                       <http://www.gnu.org/licenses/>.
+ *
+ *  : 5 $
+ *
+ ************************************************************************
+ */
+
+package org.opencadc.inventory.permissions;
```
[the remark this answers is not in the document]
Comment (jburke-cadc) [reply cadc-storage-permissions/src/main/java/org/opencadc/inventory/permissions/PermissionsAction.java]: Moved.

File: cadc-storage-permissions/src/main/java/org/opencadc/inventory/permissions/PermissionsConfig.java [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+ *  with OpenCADC.  If not, see          OpenCADC ; si ce n’est
+ *  <http://www.gnu.org/licenses/>.      pas le cas, consultez :
+ *                                       <http://www.gnu.org/licenses/>.
+ *
+ *  $Revision: 5 $
+ *
+ ************************************************************************
+ */
+
+package org.opencadc.inventory.permissions;
```
[the remark this answers is not in the document]
Comment (jburke-cadc) [reply cadc-storage-permissions/src/main/java/org/opencadc/inventory/permissions/PermissionsConfig.java]: Moved.

File: cadc-storage-permissions/src/main/java/org/opencadc/inventory/permissions/PermissionsConfig.java [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.log4j.Logger;
+import org.opencadc.gms.GroupURI;
+
```
[the remark this answers is not in the document]
Comment (jburke-cadc) [reply cadc-storage-permissions/src/main/java/org/opencadc/inventory/permissions/PermissionsConfig.java]: JavaDoc added and moved too org.opencadc.baldur.

File: cadc-storage-permissions/src/main/java/org/opencadc/inventory/permissions/xml/GrantReader.java (hunk 1 above)
Comment (jburke-cadc) [reply cadc-storage-permissions/src/main/java/org/opencadc/inventory/permissions/xml/GrantReader.java]: Named changed.

File: baldur/src/main/webapp/WEB-INF/web.xml [hunk 10]
```diff
@@ -50,6 +51,25 @@
...
     </servlet>
 
+    <!-- Permissions Servlet -->
+    <servlet>
+        <servlet-name>PermissionsServlet</servlet-name>
+        <servlet-class>ca.nrc.cadc.rest.RestServlet</servlet-class>
+        <init-param>
+            <param-name>get</param-name>
+            <param-value>org.opencadc.inventory.permissions.GetAction</param-value>
+        </init-param>
```
[the remark this answers is not in the document]
Comment (jburke-cadc) [reply baldur/src/main/webapp/WEB-INF/web.xml]: That is used by the RestServlet to load the various action classes and the cadc-rest package would require some rework.

**Comment by pdowler** [comment]
OBE pull request #48

**Pull request closed without merging**